File: golang/internal/presentation/order_controller.go
package presentation

import (
	"errors"

	"folio/codinginterview/internal/application/usecase/order"
)

type NewOrderRequest struct {
	UserId string
	Amount string
}

type AdditionalOrderRequest struct {
	UserId string
	Amount string
}

type OrderController struct {
	newOrderUsecase *order.NewOrderUsecase
	additionalBuyOrderUsecase   *order.AdditionalBuyOrderUsecase
}

func NewOrderController(
	newOrderUsecase *order.NewOrderUsecase,
	additionalBuyOrderUsecase *order.AdditionalBuyOrderUsecase,
) *OrderController {
	return &OrderController{
		newOrderUsecase: newOrderUsecase,
		additionalBuyOrderUsecase:   additionalBuyOrderUsecase,
	}
}

func (c *OrderController) NewOrder(req NewOrderRequest) error {
	uid, err := parseUserId(req.UserId)
	if err != nil {
		return err
	}
	amt, err := parseAmount(req.Amount)
	if err != nil {
		return err
	}

	err = c.newOrderUsecase.Run(order.NewOrderUsecaseInput{UserId: uid, Amount: amt})
	if err != nil {
		if errors.Is(err, order.ErrNewOrderUserAlreadyExists) {
			return newBadRequest("user already has account")
		}
		if errors.Is(err, order.ErrNewOrderAmountTooSmall) {
			return newBadRequest("amount is too small")
		}
		return err
	}
	return nil
}

func (c *OrderController) AdditionalOrder(req AdditionalOrderRequest) error {
	uid, err := parseUserId(req.UserId)
	if err != nil {
		return err
	}
	amt, err := parseAmount(req.Amount)
	if err != nil {
		return err
	}

	err = c.additionalBuyOrderUsecase.Run(order.AdditionalBuyOrderUsecaseInput{UserId: uid, Amount: amt})
	if err != nil {
		if errors.Is(err, order.ErrAdditionalBuyUserNotFound) {
			return newBadRequest("user has no live account")
		}
		if errors.Is(err, order.ErrAdditionalBuyAmountTooSmall) {
			return newBadRequest("amount is too small")
		}
		return err
	}
	return nil
}

File: java8/src/main/java/folio/codinginterview/presentation/OrderController.java
package folio.codinginterview.presentation;

import folio.codinginterview.application.usecase.order.AdditionalBuyOrderUsecase;
import folio.codinginterview.application.usecase.order.NewOrderUsecase;
import folio.codinginterview.presentation.PresentationException.BadRequestException;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;

public final class OrderController extends PresentationPreparation {
    public static final class NewOrderRequest {
        private final String userId;
        private final String amount;
        public NewOrderRequest(String userId, String amount) {
            this.userId = userId;
            this.amount = amount;
        }
        public String userId() { return userId; }
        public String amount() { return amount; }
    }

    public static final class AdditionalOrderRequest {
        private final String userId;
        private final String amount;
        public AdditionalOrderRequest(String userId, String amount) {
            this.userId = userId;
            this.amount = amount;
        }
        public String userId() { return userId; }
        public String amount() { return amount; }
    }

    private final NewOrderUsecase newOrderUsecase;
    private final AdditionalBuyOrderUsecase additionalBuyOrderUsecase;

    public OrderController(
            NewOrderUsecase newOrderUsecase,
            AdditionalBuyOrderUsecase additionalBuyOrderUsecase
    ) {
        this.newOrderUsecase = newOrderUsecase;
        this.additionalBuyOrderUsecase = additionalBuyOrderUsecase;
    }

    public CompletableFuture<Void> newOrder(NewOrderRequest req) {
        return parseUserId(req.userId()).thenCompose(uid ->
                parseAmount(req.amount()).thenCompose(amt ->
                        newOrderUsecase.run(new NewOrderUsecase.Input(uid, amt))
                                .handle((v, ex) -> {
                                    if (ex != null) {
                                        Throwable cause = ex instanceof CompletionException ? ex.getCause() : ex;
                                        if (cause instanceof NewOrderUsecase.UserAlreadyExists) {
                                            throw new CompletionException(new BadRequestException("user already has account"));
                                        }
                                        if (cause instanceof NewOrderUsecase.AmountTooSmall) {
                                            throw new CompletionException(new BadRequestException("amount is too small"));
                                        }
                                        throw new CompletionException(cause);
                                    }
                                    return null;
                                })));
    }

    public CompletableFuture<Void> additionalOrder(AdditionalOrderRequest req) {
        return parseUserId(req.userId()).thenCompose(uid ->
                parseAmount(req.amount()).thenCompose(amt ->
                        additionalBuyOrderUsecase.run(new AdditionalBuyOrderUsecase.Input(uid, amt))
                                .handle((v, ex) -> {
                                    if (ex != null) {
                                        Throwable cause = ex instanceof CompletionException ? ex.getCause() : ex;
                                        if (cause instanceof AdditionalBuyOrderUsecase.UserNotFound) {
                                            throw new CompletionException(new BadRequestException("user has no live account"));
                                        }
                                        if (cause instanceof AdditionalBuyOrderUsecase.AmountTooSmall) {
                                            throw new CompletionException(new BadRequestException("amount is too small"));
                                        }
                                        throw new CompletionException(cause);
                                    }
                                    return null;
                                })));
    }
}
